File: android/app/src/main/java/org/outreach/feature/map/MapFilterUtils.kt
package org.outreach.feature.map

import java.time.LocalDate
import java.time.format.DateTimeFormatter
import org.outreach.core.model.HouseholdRecord

fun parseIsoDateOrNull(value: String?): LocalDate? {
    if (value.isNullOrBlank()) return null
    return runCatching { LocalDate.parse(value, DateTimeFormatter.ISO_DATE) }.getOrNull()
}

fun formatBriefComment(value: String): String {
    return value
        .replace("_", " ")
        .split(" ")
        .filter { it.isNotBlank() }
        .joinToString(" ") { token ->
            token.replaceFirstChar { if (it.isLowerCase()) it.titlecase() else it.toString() }
        }
}

/** Substring match on [HouseholdRecord.name] or [HouseholdRecord.streetAddress] (case-insensitive). Blank [query] matches all. */
fun householdMatchesTextSearch(household: HouseholdRecord, query: String): Boolean {
    val q = query.trim()
    if (q.isEmpty()) return true
    return household.name.contains(q, ignoreCase = true) ||
        household.streetAddress.contains(q, ignoreCase = true)
}

fun filterHouseholdsForMap(
    data: List<HouseholdRecord>,
    visibleZipTabs: Set<String>,
    briefCommentMode: String,
    selectedBriefComments: Set<String>,
    startDate: LocalDate,
    endDate: LocalDate
): List<HouseholdRecord> {
    return data.filter { household ->
        val tabMatches =
            visibleZipTabs.isEmpty() || household.source.sheetName in visibleZipTabs
        val briefCommentMatches = when (briefCommentMode) {
            "pick_some" -> household.briefComment in selectedBriefComments
            else -> true
        }
        val visitationDate = parseIsoDateOrNull(household.lastVisited)
        val dateMatches = visitationDate == null ||
            (!visitationDate.isBefore(startDate) && !visitationDate.isAfter(endDate))
        tabMatches && briefCommentMatches && dateMatches
    }
}

fun applyOldestRecordsLimit(households: List<HouseholdRecord>, limit: Int?): List<HouseholdRecord> {
    val requested = limit ?: return households
    if (requested <= 0) return households
    return households
        .sortedWith(
            compareBy<HouseholdRecord>(
                { parseIsoDateOrNull(it.lastVisited)?.toEpochDay() ?: Long.MIN_VALUE },
                { it.id }
            )
        )
        .take(requested)
}

File: ios/Outreach/Outreach/MapFilterUtils.swift
import Foundation

func parseIsoDateOrNull(_ value: String?) -> Date? {
    guard let value, !value.isEmpty else { return nil }
    let f = ISO8601DateFormatter()
    f.formatOptions = .withFullDate
    f.timeZone = TimeZone(identifier: "UTC")
    return f.date(from: value)
}

func formatBriefComment(_ value: String) -> String {
    value
        .replacingOccurrences(of: "_", with: " ")
        .split(separator: " ")
        .filter { !$0.isEmpty }
        .map { s in
            let str = String(s)
            guard let c = str.first else { return str }
            return String(c).uppercased() + str.dropFirst().lowercased()
        }
        .joined(separator: " ")
}

func householdMatchesTextSearch(_ household: HouseholdRecord, query: String) -> Bool {
    let q = query.trimmingCharacters(in: .whitespacesAndNewlines)
    if q.isEmpty { return true }
    return household.name.range(of: q, options: .caseInsensitive) != nil
        || household.streetAddress.range(of: q, options: .caseInsensitive) != nil
}

func filterHouseholdsForMap(
    data: [HouseholdRecord],
    visibleZipTabs: Set<String>,
    briefCommentMode: String,
    selectedBriefComments: Set<String>,
    startDate: Date,
    endDate: Date
) -> [HouseholdRecord] {
    let cal = Calendar(identifier: .gregorian)
    return data.filter { household in
        let tabMatches = visibleZipTabs.isEmpty || visibleZipTabs.contains(household.source.sheetName)
        let briefCommentMatches: Bool = {
            if briefCommentMode == "pick_some" {
                return selectedBriefComments.contains(household.briefComment)
            }
            return true
        }()
        let visitationDate = parseIsoDateOrNull(household.lastVisited)
        let dateMatches: Bool = {
            guard let v = visitationDate else { return true }
            let start = cal.startOfDay(for: startDate)
            let end = cal.startOfDay(for: endDate)
            let d = cal.startOfDay(for: v)
            return d >= start && d <= end
        }()
        return tabMatches && briefCommentMatches && dateMatches
    }
}

func applyOldestRecordsLimit(_ households: [HouseholdRecord], limit: Int?) -> [HouseholdRecord] {
    guard let limit, limit > 0 else { return households }
    return households
        .sorted { a, b in
            let ad = parseIsoDateOrNull(a.lastVisited)?.timeIntervalSince1970 ?? .leastNonzeroMagnitude
            let bd = parseIsoDateOrNull(b.lastVisited)?.timeIntervalSince1970 ?? .leastNonzeroMagnitude
            if ad != bd { return ad < bd }
            return a.id < b.id
        }
        .prefix(limit)
        .map { $0 }
}
